contracts/chain-adapters/Linea_Adapter.sol

```diff
@@ -0,0 +1,135 @@
+     * @param _token The address of the token to be bridged.
+     * @param _amount The amount of the token to be bridged.
+     * @param _recipient The address that will receive the tokens on the other chain.
+     */
+    function bridgeToken(
+        address _token,
+        uint256 _amount,
+        address _recipient
+    ) external payable;
+}
+
+interface IUSDCBridge {
+    function usdc() external view returns (address);
+
+    /**
+     * @dev Sends the sender's USDC from L1 to the recipient on L2, locks the USDC sent
+     * in this contract and sends a message to the message bridge
+     * contract to mint the equivalent USDC on L2
+     * @param amount The amount of USDC to send
+     * @param to The recipient's address to receive the funds
+     */
+    function depositTo(uint256 amount, address to) external payable;
+}
+
+// solhint-disable-next-line contract-name-camelcase
+contract Linea_Adapter is AdapterInterface {
+    using SafeERC20 for IERC20;
+
+    WETH9Interface public immutable l1Weth;
+
+    IMessageService public immutable l1MessageService;
+    ITokenBridge public immutable l1TokenBridge;
+    IUSDCBridge public immutable l1UsdcBridge;
+
+    /**
+     * @notice Constructs new Adapter.
+     * @param _l1Weth WETH address on L1.
+     * @param _l1MessageService Canonical message service contract on L1.
+     * @param _l1TokenBridge Canonical token bridge contract on L1.
+     * @param _l1UsdcBridge L1 USDC Bridge to ConsenSys's L2 Linea.
+     */
+    constructor(
+        WETH9Interface _l1Weth,
+        IMessageService _l1MessageService,
+        ITokenBridge _l1TokenBridge,
+        IUSDCBridge _l1UsdcBridge
+    ) {
+        l1Weth = _l1Weth;
+        l1MessageService = _l1MessageService;
+        l1TokenBridge = _l1TokenBridge;
+        l1UsdcBridge = _l1UsdcBridge;
+    }
+
+    /**
+     * @notice Send cross-chain message to target on Linea.
+     * @param target Contract on Linea that will receive message.
+     * @param message Data to send to target.
+     */
+    function relayMessage(address target, bytes calldata message) external payable override {
+        l1MessageService.sendMessage{ value: msg.value }(target, 0, message);
+        emit MessageRelayed(target, message);
+    }
+
+    /**
+     * @notice Bridge tokens to Linea.
+     * @param l1Token L1 token to deposit.
+     * @param l2Token L2 token to receive.
+     * @param amount Amount of L1 tokens to deposit and L2 tokens to receive.
+     * @param to Bridge recipient.
+     */
+    function relayTokens(
+        address l1Token,
+        address l2Token,
+        uint256 amount,
+        address to
+    ) external payable override {
+        // If the l1Token is WETH then unwrap it to ETH then send the ETH directly
+        // via the Canoncial Message Service.
+        if (l1Token == address(l1Weth)) {
+            l1Weth.withdraw(amount);
+            l1MessageService.sendMessage{ value: amount }(to, 0, "");
```

Comment:
Because Linea (currently at least) does not support auto-claiming messages containing calldata, we need to finalize most messages ourselves. In this case, though, we could set a fee because the calldata is empty. But as we need to finalize all other messages anyways, I am wondering if it makes sense to do it or not 🤔

Reply:
i think its ideal if we can get someone else to finalize this for us if we can set the fee correctly